bftengine/tests/controllerWithSimpleHistory/ControllerWithSimpleHistory_test.cpp

```diff
@@ -145,7 +147,319 @@ TEST(ControllerWithSimpleHistory, onNewView_not_primary) {
+  const std::string getVersion() const { return "v"; }
+  void serializeDataMembers(std::ostream &) const {}
+  void deserializeDataMembers(std::istream &) {}
+  const IShareSecretKey &getShareSecretKey() const { return is; }
+};
+
+// Test - Downgrade from OPTIMISTIC_FAST to FAST_WITH_THRESHOLD then upgrade to OPTIMISTIC_FAST
+// E.L Logic: Not Possible, Bug ?
+
+// Test - Downgrade from OPTIMISTIC_FAST to SLOW then OPTIMISTIC_FAST
+// Logic:
+// 1)Construct controller with C == 0 i.e. disable FAST_WITH_THRESHOLD path.
+// 2)Run loop for EvaluationPeriod times:
+//  - call to onStartingSlowCommit for more than factor * EvaluationPeriod.
+// 3)Run loop for EvaluationPeriod times:
+//  - call onMessage for each replica to satisfy full cooporation.
+TEST(ControllerWithSimpleHistory, upgrade_slow_to_optimistic) {
+  uint16_t C = 0;
+  uint16_t F = 1;
+  ReplicaId replicaId = 0;
+  ViewNum initialView = 0;
+  SeqNum initialSeq = 0;
+  ControllerWithSimpleHistory cwsh{C, F, replicaId, initialView, initialSeq};
+
+  ASSERT_EQ(CommitPath::OPTIMISTIC_FAST, cwsh.getCurrentFirstPath());
+
+  // Factor + 1, in order to trigger the degradation.
+  auto slowStartCount =
+      ControllerWithSimpleHistory_debugDowngradeFactor * ControllerWithSimpleHistory::EvaluationPeriod + 1;
+  bool changed{false};
+  auto prePrepareTime = std::chrono::steady_clock::now() - std::chrono::milliseconds(10);
+  for (auto i = (size_t)1; i <= ControllerWithSimpleHistory::EvaluationPeriod; ++i) {
+    cwsh.onSendingPrePrepare((SeqNum)i, CommitPath::OPTIMISTIC_FAST, prePrepareTime);
+    changed = cwsh.onNewSeqNumberExecution((SeqNum)i);
+    if (slowStartCount-- > 0) cwsh.onStartingSlowCommit((SeqNum)i);
+  }
+
+  ASSERT_EQ(true, changed);
+  ASSERT_EQ(CommitPath::SLOW, cwsh.getCurrentFirstPath());
+
+  char buf[5] = {'m', 'o', 's', 'h', 'e'};
+  Digest d{buf, 5};
+  ThreshSigMock th;
+  for (auto i = (size_t)65; i < 65 + ControllerWithSimpleHistory::EvaluationPeriod; ++i) {
+    cwsh.onSendingPrePrepare((SeqNum)i, CommitPath::OPTIMISTIC_FAST, prePrepareTime);
+    changed = cwsh.onNewSeqNumberExecution((SeqNum)i);
+    for (auto id : {1, 2, 3}) {
+      auto p = impl::PreparePartialMsg::create(0, i, id, d, &th);
+      cwsh.onMessage(p);
+      delete p;
+    }
+  }
+
+  ASSERT_EQ(true, changed);
+  ASSERT_EQ(CommitPath::OPTIMISTIC_FAST, cwsh.getCurrentFirstPath());
+}
+
+// Test - Downgrade from OPTIMISTIC_FAST to FAST_WITH_THRESHOLD to SLOW then FAST_WITH_THRESHOLD
+// Logic:
+// 1)Construct controller with C  > 0 i.e. enable FAST_WITH_THRESHOLD path.
+// 2)Run loop for EvaluationPeriod times:
+//  - call to onStartingSlowCommit for more than factor * EvaluationPeriod.
+// 3)Run loop for EvaluationPeriod times:
+//  - call to onStartingSlowCommit for more than factor * EvaluationPeriod.
+// 4)Run loop for EvaluationPeriod times:
+//  - call onMessage for 3*f+C (partial set) replicas to satisfy full cooporation.
```

Comment:
I wonder why the partial set size is `3f+c`, instead of `3f+c+1`, given that `n-c=3f+c+1`

Reply:
That is how it's defined in the code
` for (SeqNum i = minSeq; i <= maxSeq; i++) {
      SeqNoInfo& s = recentActivity.get(i);
      if (s.replicas.size() == (numOfReplicas - 1))
        cyclesWithFullCooperation++;
      else if (s.replicas.size() >= (3 * f + c))
        cyclesWithPartialCooperation++;
    }`